From 3dfd139c952856a00abc678f7aaeb330f7ebc2f2 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Tue, 19 Aug 2014 19:30:55 -0400
Subject: [scsi] qla4xxx: Use kmemdup instead of kmalloc + memcpy

Message-id: <1408476658-5056-18-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88345
O-Subject: [PATCH RHEL 7.1 e-stor 17/20] qla4xxx: Use kmemdup instead of kmalloc + memcpy
Bugzilla: 1089349
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Benoit Taine <benoit.taine@lip6.fr>

scsi commit id 3d30b439f0f02416e8bd784a52d0321e0c25f9fc

This issue was reported by coccicheck using the semantic patch
at scripts/coccinelle/api/memdup.cocci

Signed-off-by: Benoit Taine <benoit.taine@lip6.fr>
Acked-by: Vikas Chaudhary <vikas.chaudhary@qlogic.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla4xxx/ql4_os.c b/drivers/scsi/qla4xxx/ql4_os.c
index 23fe892..1814eaa 100644
--- a/drivers/scsi/qla4xxx/ql4_os.c
+++ b/drivers/scsi/qla4xxx/ql4_os.c
@@ -3565,14 +3565,13 @@ static int qla4xxx_copy_from_fwddb_param(struct iscsi_bus_flash_session *sess,
  if (test_bit(OPT_IPV6_DEVICE, &options)) {
   conn->ipv6_traffic_class = fw_ddb_entry->ipv4_tos;
 
-  conn->link_local_ipv6_addr = kzalloc(IPv6_ADDR_LEN, GFP_KERNEL);
+  conn->link_local_ipv6_addr = kmemdup(
+     fw_ddb_entry->link_local_ipv6_addr,
+     IPv6_ADDR_LEN, GFP_KERNEL);
   if (!conn->link_local_ipv6_addr) {
    rc = -ENOMEM;
    goto exit_copy;
   }
-
-  memcpy(conn->link_local_ipv6_addr,
-         fw_ddb_entry->link_local_ipv6_addr, IPv6_ADDR_LEN);
  } else {
   conn->ipv4_tos = fw_ddb_entry->ipv4_tos;
  }
-- 
1.7.1